src/spearman.jl

```diff
@@ -0,0 +1,169 @@
+    println(io, ident, "adjustment for ties in y:            ", x.ytiesadj)
+end
+
+function P_from_null_S_values(S_null, x::SpearmanCorrelationTest, tail)
+    S_null_mean = mean(S_null)
+    # S is approximately normally distributed
+    # S and ρ are inversely proportional
+    S_null[:] = S_null .- S_null_mean # center
+    S_centered = x.S - S_null_mean    # center
+    if tail == :both
+        modS = abs(S_centered)
+        mean(S_null .<= -modS) + mean(S_null .>= modS)
+    elseif tail == :right
+        mean(S_null .<= S_centered)
+    elseif tail == :left
+        mean(S_null .>= S_centered)
+    else
+        throw(ArgumentError("tail=$(tail) is invalid"))
+    end
+end
+
+function spearman_P_exact(x::SpearmanCorrelationTest, tail)
+    S_null = Float64[ spearman_S(perm, x.yrank) for perm in permutations(x.xrank) ]
+    P_from_null_S_values(S_null, x, tail)
+end
+
+function spearman_P_sampling(x::SpearmanCorrelationTest, tail)
+    # 360000 samples gives an se(P) < 0.0005 for P < 0.1
+    X = copy(x.xrank)
+    S_null = Float64[ spearman_S(shuffle!(X), x.yrank) for sample in 1:360000 ]
+    P_from_null_S_values(S_null, x, tail)
+end
+
+# Use estimated mean and std for the S null distribution as in:
+#
+# Press WH, Teukolsky SA, Vetterling WT, Flannery BP.
+# Numerical recipes in C.
+# Cambridge: Cambridge university press; 1996.
+function spearman_P_estimated(x::SpearmanCorrelationTest, tail)
+    N = float(x.n)
+    a = (N^3 - N)
+    # Numerical Recipes (14.6.6)
+    S_null_mean = (a/6.) - (x.xtiesadj/12.) - (x.ytiesadj/12.)
+    # Numerical Recipes (14.6.7)
+    S_null_std  = sqrt( ((N - 1.)/36.) * (1. - (x.xtiesadj/a) ) * (1. - (x.ytiesadj/a) ) ) * N * (N + 1.)
+    zscore = (x.S - S_null_mean)/S_null_std
+    # S is approximately normally distributed
+    # S and ρ are inversely proportional
+    if tail == :both
+        cdf(Normal(), -abs(zscore)) + ccdf(Normal(), abs(zscore))
+    elseif tail == :right
+        cdf(Normal(),  zscore)
+    elseif tail == :left
+        ccdf(Normal(), zscore)
+    else
+        throw(ArgumentError("tail=$(tail) is invalid"))
+    end
+end
+
+# Using T test for n > 10 as in:
+#
+# McDonald JH.
+# Handbook of biological statistics.
+# Baltimore, MD: Sparky House Publishing; 2009 Aug.
+function spearman_P_ttest(x::SpearmanCorrelationTest, tail)
+    ρ2 = x.ρ^2
+    df = x.n-2
+    t = sqrt((df*ρ2)/(1-ρ2))
+    if tail == :both
+        cdf(TDist(df), -t) + ccdf(Normal(), t)
+    elseif tail == :right
+        ccdf(TDist(df), t)
+    elseif tail == :left
+        cdf(TDist(df),  t)
+    else
+        throw(ArgumentError("tail=$(tail) is invalid"))
+    end
+end
+
+function pvalue(x::SpearmanCorrelationTest; tail=:both, method=:estimated)
+    if x.n <= 10
+        # Exact P value using permutations
+        return( spearman_P_exact(x, tail) )
+    end
+    if method == :sampling
+        return( spearman_P_sampling(x, tail) )
```

Comment:
I suggest removing the whitespace padding to be more consistent with other code in this repo